[Stdlib] Eliminate the _UnicodeEncoding_ protocol.

DougGregor opened the pull request:
This protocol was only used to fake recursive protocol conformances;
collapse it into `_UnicodeEncoding`.

DougGregor committed:
[GSB] Explicitly add several (previously-inferred) typealiases.

This works are some associated type inference failures introduced by
the next few changes.

DougGregor committed:
[GSB] Infer equivalence of same-named nested types of the same potential archetype.

When we have two nested types of a given potential archetype that have
the same name, introduce a (quietly) inferred constraint. This is
a future-proofing step for canonicalization, for a possible future
where we no longer require all of the nested types of a given name
to be equivalent, e.g., because we have a proper disambiguation
mechanism.

DougGregor committed:
[GSB] Remove a spurious variable update. NFC

DougGregor committed:
[GSB] Eliminate self-derived nested-type-name-match constraints..

Nested-type-name-match constraints are odd because they are
effectively artifacts of the GenericSignatureBuilder's algorithm,
i.e., they occur when we have two PotentialArchetypes representing the
same type, and each of those PA's has a nested type based on the same
associated type. Because of this, nested-type-name-match constraints
have no useful requirement sources, so the normal means of detecting
self-derived constraints doesn't suffice, and we end up with
self-derived nested-type-name-match constraints eliminating the
constraints they depend on, causing spurious "redundant same-type
constraint" diagnostics and minimized generic signatures that drop
important requirements.

Handle nested-type-name-match constraints by first keeping them out of
the connected-components algorithm used to compute same-type
constraints within an equivalence class. Then, detect self-derived
nested-type-name-match constraints by determining whether any of the
ancestor potential archetypes are in the same equivalence class... and
redundant with the edge that makes the ancestor potential archetypes
equivalent. Remove such self-derived edges, and treat all other
nested-type-name-match edges as derived, providing a minimized
signature.

Fixes SR-5601 and SR-5610.

DougGregor committed:
[Sequence] Make Sequence.SubSequence conform to Sequence.

Addressed ABI FIXME’s #104 and #105, making Sequence’s SubSequence conform to Sequence, with the same element type, and for which the SubSequence of a SubSequence is the same SubSequence.

DougGregor committed:
[GSB] Split out the operation to expand the requirement signature of a conformance.

NFC for now; we are going to re-use this code.

DougGregor committed:
[GSB] Process same-type requirements needed to collapse same-type components.

FIXME: Needs more description

DougGregor committed:
[GSB] Dump delayed requirements to aid in debugging. NFC

DougGregor committed:
[GSB] Collapse same-type components using delayed requirements.

DougGregor committed:
FIXUP compute remainingComponents properly.

DougGregor committed:
[GSB] Collapse structurally-equivalent potential archetypes.

FIXME HACK

DougGregor committed:
FIXUP, HACK?: Update code-completion output.

DougGregor committed:
[Stdlib] Eliminate the _UnicodeEncoding_ protocol.

This protocol was only used to fake recursive protocol conformances;
collapse it into _UnicodeEncoding.

DougGregor commented:
This builds more on top of https://github.com/apple/swift/pull/11751.

DougGregor commented:
@swift-ci please test Linux

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor committed:
Revert "FIXUP, HACK?: Update code-completion output."

This reverts commit 986df37b799f356f616c71d008608dc30cf80ad3.

DougGregor commented:
@swift-ci please test source compatibility

moiseev commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -4963,18 +5146,380 @@ namespace {
+              std::vector<IntercomponentEdge> &sameTypeEdges) {
+  SmallVector<unsigned, 4> collapsedParents;
+  for (unsigned i : indices(equivClass->derivedSameTypeComponents))
+    collapsedParents.push_back(i);
```
Aren't implicit blocks considered evil?

DougGregor replied:
I, personally, don't consider them evil... am I committing a sin?

moiseev replied:
I'm just thinking about the "gotofail".

jrose-apple replied:
I swear it used to be in the LLVM conventions, but now I don't see it and didn't see an obvious commit that removed it.

DougGregor replied:
I'll just fix it

DougGregor committed:
[Conformance checking] Find the "best" associated type default to use.

When looking for an associated type default, look at the other
protocols to which the nominal type conforms, and pick the
most-specific associated type default from the protocol hierarchy.

DougGregor committed:
Revert "[GSB] Explicitly add several (previously-inferred) typealiases."

This reverts commit e4d50d55a3b8c2a953d6a573464bdb0c33753068.

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test source compatibility

DougGregor committed:
FIXME no gotofail

DougGregor committed:
Revert "[Conformance checking] Find the "best" associated type default to use."

This reverts commit 49706334e603b781dfb84a3ec625dbe0617da62d.

DougGregor committed:
“Sanitize” requirements before checking inferred type witnesses.

When we check type inferred type witnesses against the requirement
signature of the protocol, first "sanitize" the requirements to remap
any dependent member types into the equivalent dependent member types
within that protocol, dropping any requirements that can't be
remapped. This eliminates some extraneous recursion introduced by
checking requirements, and isolates us somewhat from the specific
choice of "canonical" associated types.

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test

DougGregor commented:
Yippee! I'm going to fold this into https://github.com/apple/swift/pull/11852, because I have too many interrelated PRs going on and this isolated standard library change isn't a user-visible thing anyway.
